src/librustc_passes/rvalue_promotion.rs

```diff
@@ -104,38 +107,51 @@ struct CheckCrateVisitor<'a, 'tcx: 'a> {
     tcx: TyCtxt<'a, 'tcx, 'tcx>,
     in_fn: bool,
     in_static: bool,
-    promotable: bool,
     mut_rvalue_borrows: NodeSet,
     param_env: ty::ParamEnv<'tcx>,
     identity_substs: &'tcx Substs<'tcx>,
     tables: &'a ty::TypeckTables<'tcx>,
     result: ItemLocalSet,
 }
 
+struct Promotable;
+
+struct NotPromotable;
+
 impl<'a, 'gcx> CheckCrateVisitor<'a, 'gcx> {
     // Returns true iff all the values of the type are promotable.
     fn type_has_only_promotable_values(&mut self, ty: Ty<'gcx>) -> bool {
+        debug!("In type_has_only_promotable_values(), ty: {:#?}", ty);
+
         ty.is_freeze(self.tcx, self.param_env, DUMMY_SP) &&
             !ty.needs_drop(self.tcx, self.param_env)
     }
 
-    fn handle_const_fn_call(&mut self, def_id: DefId, ret_ty: Ty<'gcx>, span: Span) {
-        self.promotable &= self.type_has_only_promotable_values(ret_ty);
+    fn handle_const_fn_call(&mut self, def_id: DefId,
+                            ret_ty: Ty<'gcx>, span: Span) -> Result<Promotable, NotPromotable> {
+        debug!("In handle_const_fn_call(), def_id: {:#?}, span: {:#?}, ret_ty: {:#?}",
+               def_id, span, ret_ty);
```

Comment:
Same here. Also, please don't use `{:#?}` unless necessary, it makes debug logs harder to work with. Also, for `Ty` you only need `{}`, not even `{:?}`.

Reply:
That one's on me. The logs were unreadable in the unexpanded version (well not for types, but everything else) I wish we had a way to switch the style with an env var

Reply:
Yea the ASTs. They were pretty helpful during debugging.